src/briefcase/integrations/docker.py

```diff
@@ -117,136 +174,131 @@ def verify_docker(command):
                 raise BriefcaseCommandError(
                     WRONG_DOCKER_VERSION_ERROR.format(
                         docker_version=docker_version,
-                        **docker_install_details(command.host_os),
+                        **docker_install_details(tools.host_os),
                     )
                 )
 
         else:
-            command.logger.warning(UNKNOWN_DOCKER_VERSION_WARNING)
+            tools.logger.warning(UNKNOWN_DOCKER_VERSION_WARNING)
     except subprocess.CalledProcessError:
-        command.logger.warning(DOCKER_INSTALLATION_STATUS_UNKNOWN_WARNING)
+        tools.logger.warning(DOCKER_INSTALLATION_STATUS_UNKNOWN_WARNING)
     except FileNotFoundError as e:
         # Docker executable doesn't exist.
         raise BriefcaseCommandError(
-            DOCKER_NOT_INSTALLED_ERROR.format(**docker_install_details(command.host_os))
+            DOCKER_NOT_INSTALLED_ERROR.format(**docker_install_details(tools.host_os))
         ) from e
 
     # Check that docker commands can actually run.
-    _verify_docker_can_run(command)
-
-    # Return the Docker wrapper
-    return Docker
-
-
-def _verify_docker_can_run(command):
-    """Verify that the user has sufficient permissions, and that the Docker
-    daemon is running.
-
-    This is done by attempting to run a low-impact command (docker info)
-    that requires both permissions and a working daemon.
-
-    :param command: The command that needs to perform the verification check.
+    _verify_docker_can_run(tools)
+
+    tools.docker = True
+
+
+def verify_docker_for_app(
```

Comment:
This contains two compromises.

Since `subprocess` is innately set on `tools`, it is necessary to verify its type here as part of the short circuit. This may be necessary, in general. for tools that override another tool. I originally was verifying `Subprocess` outside of `ToolCache` but this started to become problematic because several codepaths assume `subprocess` is available....so, it started to make more sense to put it on `tools` by default instead of trying to verify it everywhere.

It is necessary to call `Docker().prepare()` here instead of in `AppImage.verify_app_tools()` to avoid also performing the short circuiting checks there. If the short circuit checks are skipped altogether, we lose part of the original impetus for this PR: ensuring the Docker image exists and is up to date.....but only doing that once.

Reply:
This reasoning makes sense; but I wonder if we might be better served by addressing the underlying complexity - the fact that a tool is overwriting a different tool.

I wonder if we could clean this up by acknowledging this fact. Split `Docker` into two parts - the Docker tool, which is a pure, app-agnostic representation of "Docker exists"; and an `AppContext` class that is a separate "tool" from the perspective of the tool cache, but is represents the binding of a single app to the Docker instance. (This might not even need a separate class - just recognise that the current Docker class is really an AppContext. The base "Docker" tool wouldn't be much more than a placeholder for verification logic; although it could be a place to put "list docker images" type functionality in future).

There would be 2 versions of an `AppContext` class - a `DockerAppContext` whose verification prepares the Docker container for the app, and is bound to that specific app; and a `NativeAppContext` (part of the subprocess module) whose verification just copies a reference to `tools.Subprocess`. Both implementations would bind to `tools[app].app_context`.

This would allow `verify_tools` to verify the existence of Docker, and for `tools.docker` to be an actual Docker instance. This avoids the weird `self.docker = True` workaround; and allows `verify_app_tools` to contain the actual logic - `if use_docker: verify a DockerAppContext; else: verify a NativeAppContext`; and the individual AppContext tools only need to check if `app_context` has been verified (no need for a type check).

End users wanting to run something would invoke `tools[app].app_context.run()` - which is arguably a little clearer than `tools[app].subprocess` anyway, as it removes the ambiguity over whether we're dealing with a "real" subprocess.

Reply:
This is cleaner and I have it implemented.

This strategy does create a problem in the chain of `super()` calls. The `CreateCommand` needs `tools[app].app_context` initialized whether or not Docker is in play. So, to avoid setting `tools[app].app_context` to `Subprocess` before it can be set to `Docker`, the `super()` needs to run after `Docker` is verified. This isn't the end of the world, though.